zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java

```diff
@@ -240,16 +249,47 @@ public void addLearnerHandler(LearnerHandler learner) {
      * @param peer
      */
     @Override
-    public void removeLearnerHandler(LearnerHandler peer) {
-        synchronized (forwardingFollowers) {
-            forwardingFollowers.remove(peer);
-        }
-        synchronized (learners) {
-            learners.remove(peer);
-        }
-        synchronized (observingLearners) {
-            observingLearners.remove(peer);
-        }
+    public synchronized void removeLearnerHandler(LearnerHandler peer) {
+        forwardingFollowers.remove(peer);
```

Comment:
you have changed the synchronization. that for the removes. that needs to be fixed.

Reply:
I was trying to avoid the nested synchronized block. We need forwardingFollower and outstandingProposal in the same synchronized block. Let me see what alternatives we have...